operator/pkg/awsprovider/iam/reconciler.go

```diff
@@ -0,0 +1,210 @@
+		}
+		zap.S().Infof("[%s] Created IAM Role %v", controlPlane.ClusterName(), aws.StringValue(role.RoleName))
+	}
+	if err := role.addRoleToInstanceProfile(ctx, controlPlane); err != nil {
+		return fmt.Errorf("adding instance profile to role, %w", err)
+	}
+	for _, policy := range desiredPolicies {
+		if err := role.attachPolicy(ctx, policy, controlPlane.ClusterName()); err != nil {
+			return fmt.Errorf("attaching policies to role, %w", err)
+		}
+	}
+
+	return nil
+}
+
+func (c *Controller) Finalize(ctx context.Context, controlPlane *apis.ControlPlane) error {
+	_, err := c.iam.RemoveRoleFromInstanceProfileWithContext(ctx, &iam.RemoveRoleFromInstanceProfileInput{
+		InstanceProfileName: aws.String(KitNodeInstanceProfileNameFor(controlPlane.ClusterName())),
+		RoleName:            aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+	})
+	if err != nil && !errors.IsIAMObjectDoNotExist(err) {
+		return fmt.Errorf("removing role from instance profile, %w", err)
+	}
+	_, err = c.iam.DeleteInstanceProfileWithContext(ctx, &iam.DeleteInstanceProfileInput{
+		InstanceProfileName: aws.String(KitNodeInstanceProfileNameFor(controlPlane.ClusterName())),
+	})
+	if err != nil && !errors.IsIAMObjectDoNotExist(err) {
+		return fmt.Errorf("deleting instance profile, %w", err)
+	}
+
+	for _, policy := range desiredPolicies {
+		if _, err = c.iam.DetachRolePolicyWithContext(ctx, &iam.DetachRolePolicyInput{
+			PolicyArn: aws.String(policy),
+			RoleName:  aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+		}); err != nil {
+			return fmt.Errorf("detaching policy from role, %w", err)
+		}
+	}
+	_, err = c.iam.DeleteRoleWithContext(ctx, &iam.DeleteRoleInput{
+		RoleName: aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+	})
+	if err != nil && !errors.IsIAMObjectDoNotExist(err) {
+		return fmt.Errorf("deleting role, %w", err)
+	}
+	zap.S().Infof("[%s] Deleted IAM Role %v and instance profile",
+		controlPlane.ClusterName(), KitNodeRoleNameFor(controlPlane.ClusterName()))
+	return nil
+}
+
+func (c *Controller) getRole(ctx context.Context, controlPlane *apis.ControlPlane) (*role, error) {
+	roleOutput, err := c.iam.GetRoleWithContext(ctx, &iam.GetRoleInput{
+		RoleName: aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("getting iam role %v, %w", KitNodeRoleNameFor(controlPlane.ClusterName()), err)
+	}
+	return &role{iam: c.iam, Role: roleOutput.Role}, nil
+}
+
+func (c *Controller) createRole(ctx context.Context, controlPlane *apis.ControlPlane) (*role, error) {
+	roleOutput, err := c.iam.CreateRole(&iam.CreateRoleInput{
+		AssumeRolePolicyDocument: aws.String(assumeRolePolicyDocument),
+		Description:              aws.String("Role assumed by dataplane nodes created by KIT operated"),
+		RoleName:                 aws.String(KitNodeRoleNameFor(controlPlane.ClusterName())),
+		Tags:                     generateRoleTags(controlPlane.ClusterName()),
```

Comment:
function name sounds generic but this function is only good for `kitNodeROle` based on this line. Instead of passing `controlPlane *apis.ControlPlane` as func param, i would recommend passing `rolename` as arg to this function. Thoughts ?

Reply:
Will update the function param.